Archives/ClmFile.cpp

```diff
@@ -222,242 +222,175 @@ namespace Archives
 		}
 
 		return true;
 	}
 
 	// Searches through the wave file to find the given chunk length
-	// The current file pointer is set the the first byte after the chunk header
+	// The current stream position is set the the first byte after the chunk header
 	// Returns the chunk length if found or -1 otherwise
-	int ClmFile::FindChunk(int chunkTag, HANDLE file)
+	int ClmFile::FindChunk(uint32_t chunkTag, FileStreamReader& fileStreamReader)
 	{
 #pragma pack(push, 1)
 		struct ChunkHeader
 		{
-			int formatTag;
-			int length;
+			uint32_t formatTag;
+			uint32_t length;
 		};
 #pragma pack(pop)
 
-		unsigned long numBytesRead;
-		ChunkHeader header;
-		int curPos;
-		int fileSize;
+		uint64_t fileSize = fileStreamReader.Length();
+
+		if (fileSize < 20) {
+			return -1;
+		}
 
-		fileSize = GetFileSize(file, nullptr);
-		if (fileSize < 20) return -1;
 		// Seek to beginning of first internal chunk (provided it exists)
 		// Note: this seeks past the initial format tag (such as RIFF and WAVE)
-		if (SetFilePointer(file, 12, nullptr, FILE_BEGIN) == -1) return -1;
+		fileStreamReader.Seek(12);
 
-		curPos = 12;
+		ChunkHeader header;
+		uint32_t currentPosition = 12;
 		do
 		{
 			// Read the tag
-			ReadFile(file, &header, sizeof(header), &numBytesRead, nullptr);
+			fileStreamReader.Read((char*)&header, sizeof(header));
+			
 			// Check if this is the right header
-			if (header.formatTag == chunkTag) return header.length;
-			// Not the right header. Skip to next header
-			curPos += header.length + 8;
-			if (SetFilePointer(file, curPos, nullptr, FILE_BEGIN) == -1) return -1;
-		} while (curPos < fileSize);
+			if (header.formatTag == chunkTag) {
+				return header.length;
+			}
+			
+			// If not the right header, skip to next header
+			currentPosition += header.length + 8;
+			fileStreamReader.Seek(currentPosition);
+		} while (currentPosition < fileSize);
 
 		return -1;	// Failed to find the tag
 	}
 
-	// Closes all input files in the array fileHandle (of size numFilesToPack)
-	// and releases the memory for the file handles and the wave format structures.
-	void ClmFile::CleanUpVolumeCreate(HANDLE outFile,
-		int numFilesToPack,
-		HANDLE *fileHandle,
-		WaveFormatEx *waveFormat,
-		IndexEntry *indexEntry)
-	{
-		// Close the output file
-		if (outFile) CloseHandle(outFile);
-		// Close all open input files
-		for (int i = 0; i < numFilesToPack; i++) {
-			CloseHandle(fileHandle[i]);
-		}
-		// Free temporary memory
-		delete[] fileHandle;
-		delete[] waveFormat;
-		delete[] indexEntry;
-	}
-
-	// Compares wave format structures in the array waveFormat
+	// Compares wave format structures in the waveFormats container
 	// Returns true if they are all the same and false otherwise.
-	bool ClmFile::CompareWaveFormats(int numFilesToPack, WaveFormatEx *waveFormat)
+	bool ClmFile::CompareWaveFormats(const std::vector<WaveFormatEx>& waveFormats)
 	{
-		for (int i = 1; i < numFilesToPack; i++)
+		for (WaveFormatEx waveFormat : waveFormats)
```

Comment:
The function is returning true if all wave formats are duplicates which should not result in a bug in its current form. However, it doesn't make sense to compare the first index of WaveFormats with itself before checking the rest. Definitely a mistake on my part. I'll fix it shortly.

Reply:
Ahh, right! The check is opposite what I was thinking. I was thinking in terms of a duplicate filename check, where of course you'd expect them to all be different. In that case, maybe don't worry about this then, since the code will be subject to change if we start processing each Wave file one at a time, using delayed opening.